pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go

```diff
@@ -74,7 +82,61 @@ func (ca *cloudCIDRAllocator) PerformMultiNetworkCIDRAllocation(node *v1.Node, i
 			}
 		}
 	}
-	return defaultNwCIDRs, northInterfaces, additionalNodeNetworks, nil
+	if err = updateAnnotations(node, northInterfaces, additionalNodeNetworks); err != nil {
+		return nil, err
+	}
+	return defaultNwCIDRs, nil
+}
+
+func updateAnnotations(node *v1.Node, northInterfaces networkv1.NorthInterfacesAnnotation, additionalNodeNetworks networkv1.MultiNetworkAnnotation) error {
+	northInterfaceAnn, err := networkv1.MarshalNorthInterfacesAnnotation(northInterfaces)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal the north interfaces annotation for multi-networking", "nodeName", node.Name)
+		return err
+	}
+	additionalNodeNwAnn, err := networkv1.MarshalAnnotation(additionalNodeNetworks)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal the additional node networks annotation for multi-networking", "nodeName", node.Name)
+		return err
+	}
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	node.Annotations[networkv1.NorthInterfacesAnnotationKey] = northInterfaceAnn
+	node.Annotations[networkv1.MultiNetworkAnnotationKey] = additionalNodeNwAnn
+	node.Status.Capacity, err = allocateIPCapacity(node, additionalNodeNetworks)
+	return err
+}
+
+// allocateIPCapacity updates the extended IP resource capacity for every non-default network on the node.
+func allocateIPCapacity(node *v1.Node, nodeNetworks networkv1.MultiNetworkAnnotation) (v1.ResourceList, error) {
+	resourceList := node.Status.Capacity
+	if resourceList == nil {
+		resourceList = make(v1.ResourceList)
+	}
+	// Rebuild the IP capacity for all the networks on the node by deleting the existing IP capacities first.
+	for name := range resourceList {
+		if strings.HasPrefix(name.String(), networkv1.NetworkResourceKeyPrefix) {
+			delete(resourceList, name)
+		}
+	}
+	for _, nw := range nodeNetworks {
+		if len(nw.Cidrs) < 1 {
+			return nil, fmt.Errorf("network %s is missing CIDRs", nw.Name)
+		}
+		_, ipNet, err := net.ParseCIDR(nw.Cidrs[0])
+		if err != nil {
+			return nil, err
+		}
+		var ipCount int64 = 1
+		size := netutils.RangeSize(ipNet)
+		if size > 1 {
+			// The number of IPs supported are halved and returned for overprovisioning purposes.
+			ipCount = size >> 1
+		}
```

Comment:
you should put this into a small function and make it testable to avoid subtle bugs with off by 1 errors or int overflows
```
func ipSize(net *net.IPNet) int64 {
}
```

this is based on experience, we had these bugs depending on `netutils.RangeSize()`, it also help to document it, because in the test you can see what is the expected number of IPs for subnet X

Reply:
refactored, see the tests.